src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -0,0 +1,658 @@
+                for ref in self._defsout:
+                    self._reaches.append(ref)
+            else:
+                # If this block killed any accesses, then the first element
+                # of the killed writes is the access access that we're
+                # dependent with.
+                self._reaches.append(self.killed[0])
+
+            # We're not in a control flow region, so we stop if the
+            # reference is written to, so we don't need to ever add
+            # elements of the killed array here.
+            # FIXME This is wrong.
+
+        # Reset the start and stop points before returning the result.
+        self._start_point = save_start_position
+        self._stop_point = save_stop_position
+        return self._reaches
+
+    def _compute_forward_uses(self, basic_block_list):
+        """
+        Compute the forward uses for self._reference for the
+        basic_block_list provided. This function will not work
+        correctly if there is control flow inside the
+        basic_block_list provided.
+        Reads to the reference that occur before a write will
+        be added to the self._uses array, the final write will
+        be provided as self._defsout and all previous writes
+        will be inside self._killed.
+
+        :param basic_block_list: The list of nodes that make up the basic
+                                 block to find the forward uses in.
+        :type basic_block_list: list[:py:class:`psyclone.psyir.nodes.Node`]
+
+        :raises InternalError: If a GOTO statement is found in the code
+                               region.
+        """
+        sig, _ = self._reference.get_signature_and_indices()
+        # For a basic block we will only ever have one defsout
+        defs_out = None
+        for region in basic_block_list:
+            for reference in region.walk((Reference, Call, CodeBlock, Return)):
+                # Store the position instead of computing it twice.
+                abs_pos = reference.abs_position
+                if abs_pos <= self._start_point or abs_pos >= self._stop_point:
+                    continue
+                if isinstance(reference, Return):
+                    # When we find a return statement any following statements
+                    # can be ignored so we can return.
+                    if defs_out is not None:
+                        self._defsout.append(defs_out)
+                    return
+                if isinstance(reference, CodeBlock):
+                    # CodeBlocks only find symbols, so we can only do as good
+                    # as checking the symbol - this means we can get false
+                    # positives for structure accesses inside CodeBlocks.
+                    if isinstance(reference._fp2_nodes[0], Goto_Stmt):
+                        raise InternalError("DefinitionUseChains can't handle "
+                                            "code containing GOTO statements.")
```

Comment:
Could we use NotImplementedError (also in the docstring)?

Reply:
Yeah that seems better, fixed.